keps/sig-cloud-provider/003-testgrid-e2e.md

```diff
@@ -0,0 +1,224 @@
+  - "@andrewsykim"
+owning-sig: sig-cloud-provider
+participating-sigs:
+  - sig-testing
+  - sig-release
+  - sig-aws
+  - sig-azure
+  - sig-gcp
+  - sig-ibmcloud
+  - sig-openstack
+  - sig-vmware
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2018-06-06
+last-updated: 2018-06-06
+status: provisional
+
+---
+
+# Reporting Conformance Test Results to Testgrid
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+This is a KEP outlining the motivation behind why cloud providers should frequently upload E2E conformance test results to [Testgrid](https://github.com/kubernetes/test-infra/tree/master/testgrid) and how a cloud provider can go about doing this.
+
+## Motivation
+
+The primary motivation behind collecting conformance test results from various cloud providers on a regular basis is to inform sig-release of any critical bugs. It's important to collect results from various cloud providers to increase coverage and inform other cloud providers of bugs they may be impacted by.
+
+### Goals
+
+* All SIGs / Subprojects owners representing a cloud provider is aware of the importance of frequently uploading conformance test results to testgrid
+* There is a clear and detailed process in place for any cloud provider to upload conformance test results to Testgrid.
+
+### Non-Goals
+
+* Test coverage - increasing test coverage is outside the scope of this KEP
+* CI/CD - what CI/CD tool is used to run E2E tests and upload results is outside the scope of this KEP. It is up to the cloud provider to decide where/how tests are run.
+* Cluster Provisioning - how a cluster is provisioned is outside the scope of this KEP.
+
+## Proposal
+
+We would like to propose that every Kubernetes cloud provider reports conformance test results for every patch version of Kubernetes at the minimum. Running conformance tests against master and on pre-release versions are highly encouraged but will not be a requirement.
+
+### Implementation Details/Notes/Constraints
+
+Before continuing, it is highly recommended you read the following documentation provided by sig-testing:
+* [Testgrid Configuration](https://github.com/kubernetes/test-infra/tree/master/testgrid#testgrid)
+* [Display Conformance Tests with Testgrid](https://github.com/kubernetes/test-infra/blob/master/testgrid/conformance/README.md)
+
+#### How to Run E2E Conformance Tests
+
+[Sonobuoy](https://github.com/heptio/sonobuoy) is the recommended tool to run conformance tests on your cluster. At this point we will assume that you have a running cluster and that `kubectl` is configured to point to that cluster with admin access.
```

Comment:
There is actually a major downside to using Sonobuoy vs building the e2e tests with kubernetes in continuous integration in particular, AFAIK. If you use the pre-released Sonobuoy images then you will not be testing the latest conformance tests as they are developed. Running these directly or with [kubetest](https://github.com/kubernetes/test-infra/tree/master/kubetest) avoids this.

Reply:
+1 on this comment. In the past whenever I wished to run conformance on the master to get latest coverage or test latest changes I used kubetest. If Sonobuoy supports this it will be good to call that out here. If not, plz include instructions on how to achieve this using Kubetest or the hack scripts. Thanks.

Reply:
Sounds good, I will update to use kubetest, I was in favour of Sonobuoy originally because it was easier to run since it doesn't require setting up a development environment but I don't think the tradeoff is worth it.

Reply:
I am not suggesting to remove Sonobuoy completely from here. But plz do mention Kubetest also as a means of triggering tests, esp if you want to test on latest changes.

Reply:
Yeah, I don't want to say that we necessarily need to mention any particular tool, I've mentioned Sonobuoy and kubetest in [the testgrid conformance usage guide](https://github.com/kubernetes/test-infra/tree/master/testgrid/conformance#usage-guide).
It should also be possible to run gingko / the conformance tests directly or via the hack scripts in kubernetes, but we should probably mention at least one way that involves using freshly built e2e tests, and how to use it correctly.

Reply:
Also: I definitely hear those concerns 😬 

I'd hope providers can eventually set up a matrix covering many versions, but even in small matrices I'd hope they can set aside one job to run (latest, latest), these tests are generally quite stable, which is expected of conformance tests.

The providers ought to be able to cross reference each other if we can track upstream Kubernetes together w/o pinning, at least for one periodic "job" each. If all of the providers start failing around the same time, it's probably Kubernetes, not the provider.

Reply:
This will be one of the top concerns for sure. I don't think we need a full proof solution yet and I think KEPs are supposed to be updated frequently so we can iterate on the KEP once we have a better story, likely once all the providers have some experience with the workflow. I like that with `kubetest` we do give every provider the ability to choose that (cluster version is version tested against, kubernetes repo version is version for conformance tests) which I think is enough for us to move forward.

Reply:
So sonobuoy will lag for the release just due the way artifacts are published.  If we were to continuously publish an upstream conformance test similar to what we do downstream the original statement is a non-starter.  This has been a todo to publish the artifact, but it's never been top of mind.  I'd be happy to take it up on 1.12.

I'll bring this up during the next sig-testing meeting.

Reply:
Sure, will these artifacts work for each branch? 
Conformance tests in particular are pretty easy to run from ginkgo / e2e.test directly without many args but for continuous integration really ought to be in sync with the k8s binaries directly... 
My concern is that the Sonobuoy e2e binaries won't be able to match the k8s updates when testing `master` in particular, unless the test can also build sonobuoy images ..?

Building and running the e2e tests from source for master continuous integration in particular seems unavoidable, but not necessarily for actual releases.

Reply:
The [testgrid conformance docs in test-infra](https://github.com/kubernetes/test-infra/tree/master/testgrid/conformance) call out using the official Sonobuoy config at https://github.com/cncf/k8s-conformance as the first option, which should *Definitely* be the default for "am I conformant", as that's what you should submit to the CNCF.

For CI though, to test with the latest build something like:
```sh
kubetest \
# grab the most recent CI tarball of kubernetes, including the tests
    --extract=ci/latest \ 
# conformance tests aren't supposed to be aware of providers
    --provider=skeleton \ 
# tell ginkgo to only run conformance tests
    --test --test_args="--ginkgo.focus=\[Conformance\]" \  
# make sure to grab the log and junit
     --dump=./_artifacts | tee ./e2e.log 
```
should work in the meantime.

Reply:
All we need todo is push some of the details from https://github.com/heptio/kube-conformance into an upstream artifact that is cleanly factored (does just 1 thing).  Some of the test containers from upstream, last I looked, rolled in a bunch of other stuff which from a community perspective can be confusing and unnecessary.  @mml and I wanted todo this several releases ago, but just never got to it.

Reply:
Sonobuoy itself is a more elaborate runner to gather multiple artifacts in a more reproduciable manner.  You don't need to rebuild those containers and they are widely used, vetted, versioned and tested.

Reply:
there aren't any other artifacts to gather in this case...?
But you should use the latest e2es in some cases (continuous integration...) and I don't see how that will work without rebuilding https://github.com/heptio/kube-conformance/blob/c43a9d2d1f096ccff8d5b069691a523b1b517b8e/Dockerfile#L24 to match the newest binaries.

Reply:
Depends on what they want todo.  A lot of folks create their own custom tests that they collect together.  Also, it's a reproducible artifact that you can handoff.  Can't do that with the e2es today.

Reply:
xref: https://github.com/kubernetes/test-infra/issues/8299 issue for sonobuoy automation instructions with the CNCF config + the testgrid uploader.